src/util/pup_stl.h

```diff
@@ -516,7 +516,12 @@ using Requires = typename requires_impl<
   inline void pup(PUP::er &p, std::shared_ptr<T> &t) {
     PUP::able* _ = (p.isUnpacking()) ? nullptr : t.get();
     p(&_);
-    if (p.isUnpacking()) { t.reset(dynamic_cast<T*>(_)); }
+    if (p.isUnpacking()) {
+      // the shared ptr must be created with the original PUP::able ptr
+      // otherwise the dynamic cast can lead to invalid free's
+      // (it changes the pointer's address)
```

Comment:
How would that change the pointer's address?

Reply:
"With multiple inheritance, different base-class sub-objects might be at different locations within the complete object. In that case, a down-cast might have to adjust the pointer value to point to the complete object." (source: https://stackoverflow.com/a/27320572/1426075)

Reply:
I suppose better phrasing would be "changes the pointer's **value**" but since it holds an address I slipped a bit.